### ISSUE TITLE
[BUG] : NotImplementedError: Model data type and gradient accumulation data type must be equal to use ZeRO

### ISSUE THREAD

**Issue by conan1024hao** [microsoft/DeepSpeed#2569]
**Describe the bug**
I am using deepspeed to train my language models by script https://github.com/huggingface/transformers/blob/main/examples/pytorch/language-modeling/run_mlm.py.
When I was using deepspeed 0.7.5 there was no problem, but I encountered the following error after updating deepspeed to 0.7.6.
bf16 is using for training.
```
047: Traceback (most recent call last):
a0047:   File "run_mlm_with_data_wikicc.py", line 619, in <module>
a0047:     main()
a0047:   File "run_mlm_with_data_wikicc.py", line 568, in main
a0047:     train_result = trainer.train(resume_from_checkpoint=checkpoint)
a0047:   File "/bb/grandchallenge/gae50995/bigbird_env/lib64/python3.8/site-packages/transformers/trainer.py", line 1527, in train
a0047:     return inner_training_loop(
a0047:   File "/bb/grandchallenge/gae50995/bigbird_env/lib64/python3.8/site-packages/transformers/trainer.py", line 1596, in _inner_training_loop
a0047:     deepspeed_engine, optimizer, lr_scheduler = deepspeed_init(
a0047:   File "/bb/grandchallenge/gae50995/bigbird_env/lib64/python3.8/site-packages/transformers/deepspeed.py", line 344, in deepspeed_init
a0047:     deepspeed_engine, optimizer, _, lr_scheduler = deepspeed.initialize(**kwargs)
a0047:   File "/bb/grandchallenge/gae50995/bigbird_env/lib64/python3.8/site-packages/deepspeed/__init__.py", line 125, in initialize
a0047:     engine = DeepSpeedEngine(args=args,
a0047:   File "/bb/grandchallenge/gae50995/bigbird_env/lib64/python3.8/site-packages/deepspeed/runtime/engine.py", line 330, in __init__
a0047:     self._configure_optimizer(optimizer, model_parameters)
a0047:   File "/bb/grandchallenge/gae50995/bigbird_env/lib64/python3.8/site-packages/deepspeed/runtime/engine.py", line 1207, in _configure_optimizer
a0047:     optimizer_wrapper = self._do_optimizer_sanity_check(basic_optimizer)
a0047:   File "/bb/grandchallenge/gae50995/bigbird_env/lib64/python3.8/site-packages/deepspeed/runtime/engine.py", line 1133, in _do_optimizer_sanity_check
a0047:     raise NotImplementedError(
a0047: NotImplementedError: Model data type and gradient accumulation data type must be equal to use ZeRO
```
It seems the new function `get_data_type` in `deepspeed/runtime/engine.py` is strange:
```
def get_data_types(self):
        model_dtype = torch.float32
        if self.fp16_enabled():
            model_dtype = torch.float16
        elif self.bfloat16_enabled():
            model_dtype = torch.bfloat16

        if self._config.grad_accum_dtype == None:
            if model_dtype == torch.bfloat16:
                grad_accum_dtype = torch.float32
            else:
                grad_accum_dtype = model_dtype
        else:
            grad_accum_dtype = DtypeEnum(self._config.grad_accum_dtype).value

        return (model_dtype, grad_accum_dtype)
```
since `model_dtype` and `grad_accum_dtype` have to be same while `zero_enabled`.
```
if zero_enabled:
      if model_dtype != grad_accum_dtype:
          raise NotImplementedError(
              "Model data type and gradient accumulation data type must be equal to use ZeRO"
          )
```

**To Reproduce**
Our script:
```
#!/bin/bash

source /etc/profile.d/modules.sh
module load gcc/9.3.0 python/3.8 cuda/11.7 cudnn/8.6 nccl/2.14

export LD_LIBRARY_PATH=/usr/lib64:$LD_LIBRARY_PATH
source /path/to/bigbird_env/bin/activate

export PDSH_SSH_ARGS_APPEND='-p2222'

echo "PATH=${PATH}" > .deepspeed_env
echo "LD_LIBRARY_PATH=${LD_LIBRARY_PATH}" >> .deepspeed_env
echo "CUDA_HOME=$CUDA_HOME" >> .deepspeed_env

MASTER_ADDR=`hostname -I | cut -d' ' -f1`
cat $SGE_JOB_HOSTLIST > hostfile
sed -e 's/$/ slots=8/' -i hostfile 

deepspeed \
    --no_ssh_check \
    --no_local_rank \
    --hostfile hostfile \
    --master_addr ${MASTER_ADDR} \
    run_mlm.py \
        --model_type big_bird \
        --train_file train.txt \
        --tokenizer_name /path/to/tokenizer \
        --config_overrides="bos_token_id=2,eos_token_id=3,pad_token_id=0,vocab_size=32000" \
        --max_seq_length 4096 \
        --overwrite_output_dir \
        --output_dir output \
        --do_train \
        --gradient_accumulation_steps 1 \
        --per_device_train_batch_size 6 \
        --learning_rate 1e-4 \
        --max_steps 30000 \
        --logging_strategy steps \
        --save_strategy steps \
        --save_steps 1000 \
        --save_total_limit 2 \
        --bf16 \
        --deepspeed ds_config.json
```

Our ds_config:
```
{
    "fp16": {
        "enabled": "auto",
        "loss_scale": 0,
        "loss_scale_window": 1000,
        "initial_scale_power": 16,
        "hysteresis": 2,
        "min_loss_scale": 1
    },

    "bf16": {
	"enabled": "auto"
    },

    "optimizer": {
        "type": "AdamW",
        "params": {
            "lr": "auto",
            "betas": "auto",
            "eps": "auto",
            "weight_decay": "auto"
        }
    },

    "scheduler": {
        "type": "WarmupDecayLR",
        "params": {
            "total_num_steps": "auto",
            "warmup_min_lr": "auto",
            "warmup_max_lr": "auto",
            "warmup_num_steps": "auto"
        }
    },

    "zero_optimization": {
        "stage": 2,
        "offload_optimizer": {
            "device": "cpu",
            "pin_memory": true
        },
        "allgather_partitions": true,
        "allgather_bucket_size": 5e8,
        "overlap_comm": true,
        "reduce_scatter": true,
        "reduce_bucket_size": 5e8,
        "contiguous_gradients": true
    },

    "gradient_accumulation_steps": "auto",
    "gradient_clipping": "auto",
    "train_batch_size": "auto",
    "train_micro_batch_size_per_gpu": "auto"
}
```

**Expected behavior**
There is no error when using deepspeed 0.7.5.

**ds_report output**
```
--------------------------------------------------
DeepSpeed C++/CUDA extension op report
--------------------------------------------------
NOTE: Ops not installed will be just-in-time (JIT) compiled at
      runtime if needed. Op compatibility means that your system
      meet the required dependencies to JIT install the op.
--------------------------------------------------
JIT compiled ops requires ninja
ninja .................. [92m[OKAY][0m
--------------------------------------------------
op name ................ installed .. compatible
--------------------------------------------------
cpu_adam ............... [93m[NO][0m ....... [92m[OKAY][0m
cpu_adagrad ............ [93m[NO][0m ....... [92m[OKAY][0m
fused_adam ............. [93m[NO][0m ....... [92m[OKAY][0m
fused_lamb ............. [93m[NO][0m ....... [92m[OKAY][0m
[93m [WARNING] [0m please install triton==1.0.0 if you want to use sparse attention
sparse_attn ............ [93m[NO][0m ....... [93m[NO][0m
transformer ............ [93m[NO][0m ....... [92m[OKAY][0m
stochastic_transformer . [93m[NO][0m ....... [92m[OKAY][0m
async_io ............... [93m[NO][0m ....... [92m[OKAY][0m
utils .................. [93m[NO][0m ....... [92m[OKAY][0m
quantizer .............. [93m[NO][0m ....... [92m[OKAY][0m
transformer_inference .. [93m[NO][0m ....... [92m[OKAY][0m
spatial_inference ...... [93m[NO][0m ....... [92m[OKAY][0m
--------------------------------------------------
DeepSpeed general environment info:
torch install path ............... ['/path/to/env/lib/python3.8/site-packages/torch']
torch version .................... 1.13.0+cu117
torch cuda version ............... 11.7
torch hip version ................ None
nvcc version ..................... 11.7
deepspeed install path ........... ['/path/to/env/lib/python3.8/site-packages/deepspeed']
deepspeed info ................... 0.7.6, unknown, unknown
deepspeed wheel compiled w. ...... torch 1.13, cuda 11.7
```

**System info (please complete the following information):**
 - OS: CentOS Linux release 7.5.1804 (Core) 
 - GPU count and types: two machines with x8 A100s each
 - Python: 3.8.13

**Launcher context**
Launching experiments with the `deepspeed` launcher.

**Comment by tjruwase** [comment]
@conan1024hao, apologies for the code breakage. I believe this will be fixed by #2566. You can look there for the triage if interested.

**Comment by jeffra** [comment]
@conan1024hao, please re-open if this is not resolved by the PR.
